### ISSUE TITLE
Issues with vmq_queue drain 

### ISSUE THREAD

**Issue by beam-me-up** [erlio/vernemq#170]
I have a two node cluster and when i get the following kind of errors, the publish is not working. May i know why? 

```
@vmq_queue:wait_for_offline:229 got unknown sync event in wait_for_offline state {enqueue_many,[]}

@vmq_queue:drain:266 can't drain queue '{[],<<"369">>}' for [<0.11614.2>][<11781.2765.4>] due to wait_for_offline

@vmq_queue:drain:266 can't drain queue '{[],<<"369">>}' for [<0.11953.2>][<11781.3166.4>] due to draining
```

**Comment by ioolkos** [comment]
Hello @beam-me-up what is the context? did you just do a cluster leave?
The logs have to do with queue migration.

**Comment by beam-me-up** [comment]
No @ioolkos . no leave cluster, when i checked the cluster status, both the nodes are alive and true. I suspect few things 
1. The nodes are behind load balancers, the client can connect to one node and when they connect again, they may connect to another node and so the migration happens. 
2. Another scenario where two clients with same clientID is same trying to connect to both the nodes at the same time. 
I dont understand how the server behaves in such scenarios!

**Comment by ioolkos** [comment]
Ok, thanks.

1. Yes, that's exactly what (intentionally) happens.
2. If you have your clients reconnect automatically, this would create a race condition between the 2 clients throwing each other out of the cluster, initiating a queue migration, other client reconnects etc etc

Do you see disconnects within a regular interval?

**Comment by beam-me-up** [comment]
thanks @ioolkos but the publish is not just working for these clients but in the complete system. 

I just found another crash log where i m stuck  

```
2016-07-05 08:14:37 =CRASH REPORT====
  crasher:
    initial call: vmq_ranch:init/4
    pid: <0.21193.67>
    registered_name: []
    exception error: bad argument: [{ets,lookup,[vmq_status,ready],[]},{vmq_cluster,is_ready,0,[{file,"/home/vernemq/vernemq/_build/default/lib/vmq_server/src/vmq_cluster.erl"},{line,66}]},{vmq_reg_leader,register_subscriber,3,[{file,"/home/vernemq/vernemq/_build/default/lib/vmq_server/src/vmq_reg_leader.erl"},{line,44}]},{vmq_reg,register_subscriber,2,[{file,"/home/vernemq/vernemq/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,155}]},{vmq_mqtt_fsm,check_user,2,[{file,"/home/vernemq/vernemq/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,593}]},{vmq_mqtt_fsm,in,3,[{file,"/home/vernemq/vernemq/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,162}]},{vmq_mqtt_fsm,data_in,3,[{file,"/home/vernemq/vernemq/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,130}]},{vmq_ranch,handle_message,2,[{file,"/home/vernemq/vernemq/_build/default/lib/vmq_server/src/vmq_ranch.erl"},{line,152}]}]
    ancestors: [<0.2130.0>,<0.2129.0>,ranch_sup,<0.152.0>]
    messages: [{vmq_mqtt_fsm,check_keepalive}]
    links: [<0.2130.0>,#Port<0.891317>]
    dictionary: [{random_seed,{1468,4934,16731}}]
    trap_exit: true
    status: running
    heap_size: 2586
    stack_size: 27
    reductions: 916
  neighbours:
2016-07-05 08:14:37 =CRASH REPORT====
  crasher:
    initial call: vmq_ranch:init/4
    pid: <0.21208.67>
    registered_name: []
```

```
Ranch listener {{127,0,0,1},1883} had connection process started with vmq_ranch:start_link/4 at <0.21234.67> exit with reason: badarg
```

**Comment by ioolkos** [comment]
Hello @beam-me-up,
apologies for not being back earlier on this.

So, the cluster was always a 2-node cluster? there is no third stopped node around?

If you cannot publish at all in a cluster, then it thinks it's in a network partition and blocks. (if configured for full consistency).

Is there anything special in your configuration? (like allow_multiple_sessions on one node, and not on the other...)

**Comment by beam-me-up** [comment]
No @ioolkos , it is always a 2-node cluster. 
The configuration related to network on both nodes: 
``` 
trade_consistency = on
allow_multiple_sessions = off
```

**Comment by ioolkos** [comment]
Ok, let me rephrase the situation:
You have a 2-node cluster, where you cannot publish messages to existing subscriptions. There is only a limited number of possibilities: 1) cluster is partitioned 2) loadbalancer problems (connections from loadbalance to VerneMQ, LB healthchecks, timeouts 3) wrong topics etc, which is certainly not the case here.

- Can you connect/subscribe new clients?
- Is there a connection from the loadbalancer to VerneMQ? timeouts, healthchecks...
What loadbalancer do you use?

**Comment by beam-me-up** [comment]
Can you connect/subscribe new clients?
```
Yes, Connect works. I need to test again if Subscribe worked. I ll come back on this. 
```
Is there a connection from the loadbalancer to VerneMQ? timeouts, healthchecks... What loadbalancer do you use?
```
Yes, I use AWS LB. Connection timeout = 300 secs where the clients keepalive 90 secs. 
Healthcheck on the mqtt port and LB status was both the nodes were active and the traffic was routed to both nodes. 
```
I removed one of the node from the LB (two nodes are still in cluster), and i didnot experience the problem again. I shall test enabling them again and do some debugging and come back on this. 

Thank you so much for your time @ioolkos

**Comment by dergraf** [comment]
@beam-me-up any progress on this?

**Comment by larshesel** [comment]
I'll close this for now - @beam-me-up please feel free to re-open if needed - and we'd love to hear about your results on your testing and debugging efforts.
